sssd_test_framework/utils/pam.py

```diff
@@ -0,0 +1,284 @@
+
+            # Three failed login attempts
+            for i in range(3):
+                assert not client.auth.ssh.password("user-1", "bad_password")
+
+            assert not client.auth.ssh.password("user-1", "Secret123")
+
+            # Reset user lockout
+            client.pam.faillock("user-1").reset
+            assert client.auth.ssh.password("user-1", "Secret123")
+    """
+
+    def __init__(self, util: PAMUtils, file: str = "/etc/security/faillock.conf") -> None:
+        """
+        :param util: PAMUtils object
+        :type util: PAMUtils
+        :param file: Faillock configuration file
+        :type file: str, optional
+        """
+        self.util: PAMUtils = util
+        self.file: str = file
+        self.path: str = "/files" + self.file
+        self.args: str = f'--noautoload --transform "Simplevars.lns incl {self.file}"'
+        self.cmd: str = ""
+
+    def config_write(self) -> None:
+        """
+        Write faillock configuration.
+        :return: Self
+        :rtype: PAMFaillock
+        """
+        self.util.fs.backup(self.file)
+        self.util.logger.info(f"Saving Augeas changes to {self.file} on {self.util.host.hostname}")
+        self.util.host.ssh.run(f'echo -e "{self.cmd}save\n" | augtool --echo {self.args}')
+
+    def config_read(self) -> str:
+        """
+        Read faillock configuration as augeas tree.
+        :return: PAM access configuration
+        :rtype: str
+        """
+        self.util.logger.info(f"Reading Augeas tree {self.file} on {self.util.host.hostname}")
+        result = self.util.host.ssh.run(f"augtool {self.args} print {self.path}").stdout
+
+        return result
+
+    def config_delete(self, value: dict[str, str]) -> None:
+        """
+        Delete faillock configuration.
+        :param value: Configuration.
+        :type value: dict[str, str]
+        :return: None
+        """
+        if value is None:
+            raise ValueError("No data!")
+
+        self.util.logger.info(f"Deleting node in Augeas tree on {self.util.host.hostname}")
+        for k, v in value.items():
+            self.util.host.ssh.run(f"augtool {self.args} --autosave rm {self.path}/{k} {v}")
+
+    def config_set(self, value: dict[str, str]) -> None:
+        """
+        Set faillock configuration.
+        :param value: Configuration parameter(s) and value(s).
+        :type value: dict[str, str]
+        :return: None
+        """
+        if value is None:
+            raise ValueError("No data!")
+
+        for k, v in value.items():
+            self.cmd = self.cmd + f"set {self.path}/{k} {v}\n"
+
+    def user_info(self, user: str) -> str:
+        """
+        Get user faillock information.
+        :return: Output from faillock
+        :rtype: str
+        """
+        self.util.logger.info(f"Getting faillock information for {user} on {self.util.host.hostname}")
+        result = self.util.host.ssh.exec(["faillock", "--user", user], log_level=SSHLog.Error)
```

Comment:
I'd put any call to `faillock` in its own python module, as this has no relationship with PAM.

Reply:
ACK, added it to tools.py  part of the LinuxToolsUtils class.

Reply:
Yes, I would put it there too